cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/detector/MaintenanceEvent.java

```diff
@@ -24,20 +68,128 @@ public AnomalyType anomalyType() {
+    boolean hasProposalsToFix = hasProposalsToFix();
+    // Ensure that only the relevant response is cached to avoid memory pressure.
+    _optimizationResult.discardIrrelevantAndCacheJsonAndPlaintext();
+    return hasProposalsToFix;
   }
 
   @Override
   public String toString() {
-    // TODO: Add details on maintenance event.
-    return super.toString();
+    // Add details on maintenance event.
+    StringBuilder sb = new StringBuilder();
+    sb.append(String.format("{Handling %s", _maintenanceEventType));
+    if (_topicsWithRFUpdate != null) {
+      // Add summary for TOPIC_REPLICATION_FACTOR
+      sb.append(String.format(" by desired RF: [%s]", _topicsWithRFUpdate));
+    } else if (_brokers != null) {
+      // Add summary for ADD_BROKER / REMOVE_BROKER / DEMOTE_BROKER
+      sb.append(String.format(" for brokers: [%s]", _brokers));
+    }
+    sb.append("}");
+    return sb.toString();
+  }
+
+  @SuppressWarnings("unchecked")
+  protected void initBrokers(Map<String, ?> configs) {
+    _brokers = (Set<Integer>) configs.get(BROKERS_OBJECT_CONFIG);
+    if (_brokers == null || _brokers.isEmpty()) {
+      throw new IllegalArgumentException(String.format("Missing brokers for maintenance event of type %s.", _maintenanceEventType));
+    }
+  }
+
+  @SuppressWarnings("unchecked")
+  protected void initTopicsWithRFUpdate(Map<String, ?> configs) {
+    _topicsWithRFUpdate = (Map<Short, Set<String>>) configs.get(TOPICS_WITH_RF_UPDATE_CONFIG);
+    if (_topicsWithRFUpdate == null || _topicsWithRFUpdate.isEmpty()) {
+      throw new IllegalArgumentException(String.format("Missing %s to identify topics for replication factor update.",
+                                                       TOPICS_WITH_RF_UPDATE_CONFIG));
+    }
+  }
+
+  protected Map<Short, Pattern> topicPatternByReplicationFactor(Map<String, ?> configs) {
+    initTopicsWithRFUpdate(configs);
+    Map<Short, Pattern> topicPatternByReplicationFactor = new HashMap<>(_topicsWithRFUpdate.size());
+    _topicsWithRFUpdate.forEach((key, value) -> topicPatternByReplicationFactor.put(key, buildTopicRegex(value)));
+    return topicPatternByReplicationFactor;
   }
 
   @Override
   public void configure(Map<String, ?> configs) {
     super.configure(configs);
-    // TODO: Add configs for maintenance event.
+    KafkaCruiseControl kafkaCruiseControl = extractKafkaCruiseControlObjectFromConfig(configs, MAINTENANCE_EVENT);
+    KafkaCruiseControlConfig config = kafkaCruiseControl.config();
+    boolean allowCapacityEstimation = config.getBoolean(ANOMALY_DETECTION_ALLOW_CAPACITY_ESTIMATION_CONFIG);
+    boolean excludeRecentlyDemotedBrokers = config.getBoolean(SELF_HEALING_EXCLUDE_RECENTLY_DEMOTED_BROKERS_CONFIG);
+    boolean excludeRecentlyRemovedBrokers = config.getBoolean(SELF_HEALING_EXCLUDE_RECENTLY_REMOVED_BROKERS_CONFIG);
+    _optimizationResult = null;
+    _maintenanceEventType = (MaintenanceEventType) configs.get(MAINTENANCE_EVENT_TYPE_CONFIG);
```

Comment:
I am a little bit confused about using a config to determine the event type. Is the config per-set when cruise control is deployed or is it set in runtime? From #1248, I thought the point of having the maintenance event is to read the event from some store (i.e. a topic) and trigger actions based on the event type so I guess the next PR for the event reader will introduce the implementation to read form a topic and set this config to trigger the goal-based operation, is it correct?

Reply:
> Is the config per-set when cruise control is deployed or is it set in runtime?

This is a runtime config similar to the ones used for configuring other anomalies -- e.g. `FAILED_BROKERS_OBJECT_CONFIG` for `BrokerFailures` (see [setting the config](https://github.com/linkedin/cruise-control/blob/726397e93ae7c62f080b56bca963c60d52b60fe4/cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/detector/BrokerFailureDetector.java#L192) and [using it](https://github.com/linkedin/cruise-control/blob/976b0cd19b3ce76a80e21f583375545bde709df2/cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/detector/BrokerFailures.java#L100)). For examples of setting and using the runtime config `MAINTENANCE_EVENT_TYPE_CONFIG `, please see (1) [setting](https://github.com/linkedin/cruise-control/pull/1324/files#diff-ddbea1e8898a093f01b712c009b21bd4R101) and (2) [usage](https://github.com/linkedin/cruise-control/pull/1324/files#diff-848dffb3d09e6c493b4b2b9cf992d354R129).

> so I guess the next PR for the event reader will introduce the implementation to read form a topic and set this config to trigger the goal-based operation, is it correct?

That is correct.

Reply:
Thanks for the explanation.